tests/phpunit/tests/html/wpHtmlTagProcessor.php

```diff
@@ -0,0 +1,1960 @@
+			'Unexpected SCRIPT closer after DIV'  => array( 'console.log("<div target>")</script><div><div>' ),
+			'Unexpected SCRIPT closer before DIV' => array( 'console.log("<span>")</script><div target><div>' ),
+			'Missing SCRIPT closer'               => array( '<script>console.log("<div>");<div><div></script><div target>' ),
+			'TITLE before DIV'                    => array( '<title><div></title><div target><div>' ),
+			'SCRIPT inside TITLE'                 => array( '<title><script><div></title><div target><div></script><div>' ),
+			'TITLE in TEXTAREA'                   => array( '<textarea><div><title><div></textarea><div target></title><div>' ),
+		);
+	}
+
+	/**
+	 * @ticket 56299
+	 *
+	 * @covers WP_HTML_Tag_Processor::next_tag
+	 * @covers WP_HTML_Tag_Processor::set_attribute
+	 */
+	public function test_can_query_and_update_wrongly_nested_tags() {
+		$p = new WP_HTML_Tag_Processor(
+			'<span>123<p>456</span>789</p>'
+		);
+		$p->next_tag( 'span' );
+		$p->set_attribute( 'class', 'span-class' );
+		$p->next_tag( 'p' );
+		$p->set_attribute( 'class', 'p-class' );
+		$this->assertSame(
+			'<span class="span-class">123<p class="p-class">456</span>789</p>',
+			$p->get_updated_html()
+		);
+	}
+
+	/**
+	 * @ticket 56299
+	 *
+	 * @covers WP_HTML_Tag_Processor::next_tag
+	 * @covers WP_HTML_Tag_Processor::remove_attribute
+	 */
+	public function test_removing_specific_attributes_in_malformed_html() {
+		$p = new WP_HTML_Tag_Processor( self::HTML_MALFORMED );
+		$p->next_tag( 'span' );
+		$p->remove_attribute( 'Notifications<' );
+		$this->assertSame(
+			'<div><span class="d-md-none" /span><span class="d-none d-md-inline">Back to notifications</span></div>',
+			$p->get_updated_html()
+		);
+	}
+
+	/**
+	 * @ticket 56299
+	 *
+	 * @covers WP_HTML_Tag_Processor::set_attribute
+	 */
+	public function test_updating_specific_attributes_in_malformed_html() {
+		$p = new WP_HTML_Tag_Processor( self::HTML_MALFORMED );
+		$p->next_tag( 'span' );
+		$p->set_attribute( 'id', 'first' );
+		$p->next_tag( 'span' );
+		$p->set_attribute( 'id', 'second' );
+		$this->assertSame(
+			'<div><span id="first" class="d-md-none" Notifications</span><span id="second" class="d-none d-md-inline">Back to notifications</span></div>',
+			$p->get_updated_html()
+		);
+	}
+
+	/**
+	 * @ticket 56299
+	 *
+	 * @covers WP_HTML_Tag_Processor::add_class
+	 * @covers WP_HTML_Tag_Processor::set_attribute
+	 *
+	 * @dataProvider data_updating_attributes
+	 *
+	 * @param string $html     HTML to process.
+	 * @param string $expected Expected updated HTML.
+	 */
+	public function test_updating_attributes( $html, $expected ) {
+		$p = new WP_HTML_Tag_Processor( $html );
+		$p->next_tag();
+		$p->set_attribute( 'foo', 'bar' );
+		$p->add_class( 'firstTag' );
+		$p->next_tag();
+		$p->add_class( 'secondTag' );
+
+		$this->assertSame(
+			$expected,
+			$p->get_updated_html()
+		);
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array[]
```

Comment:
```suggestion
	 * @return array[] Test cases for the test_updating_attributes method.
```

Reply:
@adamziel See [this discussion](https://github.com/WordPress/wordpress-develop/pull/3920#discussion_r1092021101) regarding data provider docblocks.

Reply:
As noted in the others, I'll close this as resolved to avoid confusion.